test/state/mpt.cpp

```diff
@@ -0,0 +1,286 @@
+    {
+        assert(pos <= length);
+        Path p{{}};
+        p.length = length - pos;
+        std::copy_n(&nibbles[pos], p.length, p.nibbles);
+        return p;
+    }
+
+    [[nodiscard]] Path head(size_t size) const noexcept
+    {
+        assert(size < length);
+        Path p{{}};
+        p.length = size;
+        std::copy_n(nibbles, size, p.nibbles);
+        return p;
+    }
+
+    [[nodiscard]] bytes encode(bool extended) const
+    {
+        bytes bs;
+        const auto is_even = length % 2 == 0;
+        if (is_even)
+            bs.push_back(0x00);
+        else
+            bs.push_back(0x10 | nibbles[0]);
+        for (size_t i = is_even ? 0 : 1; i < length; ++i)
+        {
+            const auto h = nibbles[i++];
+            const auto l = nibbles[i];
+            assert(h <= 0x0f);
+            assert(l <= 0x0f);
+            bs.push_back(uint8_t((h << 4) | l));
+        }
+        if (!extended)
+            bs[0] |= 0x20;
+        return bs;
+    }
+};
+}  // namespace
+
+/// The MPT Node.
+///
+/// The implementation is based on StackTrie from go-ethereum.
+class MPTNode
+{
+    enum class Kind : uint8_t
+    {
+        leaf,
+        ext,
+        branch
+    };
+
+    static constexpr size_t num_children = 16;
+
+    Kind m_kind = Kind::leaf;
+    Path m_path{{}};
+    bytes m_value;
+    std::unique_ptr<MPTNode> children[num_children];
+
+    MPTNode(Kind kind, const Path& path, bytes&& value = {}) noexcept
+      : m_kind{kind}, m_path{path}, m_value{std::move(value)}
+    {}
+
+    /// Named constructor for an extended node.
+    static MPTNode ext(const Path& path, std::unique_ptr<MPTNode> child) noexcept
+    {
+        MPTNode node{Kind::ext, path};
+        node.children[0] = std::move(child);
+        return node;
+    }
+
+    /// Finds the position at witch two paths differ.
+    static size_t mismatch(const Path& p1, const Path& p2) noexcept
+    {
+        assert(p1.length <= p2.length);
+        return static_cast<size_t>(
+            std::mismatch(p1.nibbles, p1.nibbles + p1.length, p2.nibbles).first - p1.nibbles);
+    }
+
+public:
+    MPTNode() = default;
+
+    /// Named constructor for a leaf node.
+    static MPTNode leaf(const Path& path, bytes&& value) noexcept
+    {
+        return {Kind::leaf, path, std::move(value)};
+    }
+
+    void insert(const Path& path, bytes&& value);
+
+    [[nodiscard]] hash256 hash() const;
+};
+
+void MPTNode::insert(const Path& path, bytes&& value)  // NOLINT(misc-no-recursion)
+{
+    switch (m_kind)
+    {
+    case Kind::branch:
+    {
+        assert(m_path.length == 0);
+        const auto idx = path.nibbles[0];
+        auto& child = children[idx];
+        if (!child)
+            child = std::make_unique<MPTNode>(leaf(path.tail(1), std::move(value)));
+        else
+            child->insert(path.tail(1), std::move(value));
+        break;
+    }
+
+    case Kind::ext:
+    {
+        const auto m = mismatch(m_path, path);
```

Comment:
too long and complicated code to use single-letter variables imho
```suggestion
        const auto mismatch_pos = mismatch(m_path, path);
```